bionemo-recipes/models/esm2/tests/test_cp.py

```diff
@@ -0,0 +1,396 @@
+    torch.distributed.all_gather(losses_list, outputs_cp.loss, group=cp_group)
+
+    if cp_rank == 0:
+        average_cp_loss = torch.mean(torch.stack(losses_list))
+        # The average of per-rank losses should be close to the non-distributed loss
+        # Note: They may not be exactly equal due to how loss is computed on sharded data
+        torch.testing.assert_close(
+            average_cp_loss.cpu(),
+            loss_nondistributed_for_comparison,
+            atol=0.1,  # Allow some difference due to loss computation on shards
+            rtol=0.05,
+        )
+
+    # Gather the logits from all CP ranks
+    # The logits are split along the sequence dimension (dim=1 for THD format: [batch, seq, vocab])
+    logits_contiguous = outputs_cp.logits.contiguous()
+    logits_list = [torch.zeros_like(logits_contiguous) for _ in range(cp_world_size)]
+    torch.distributed.all_gather(logits_list, logits_contiguous, group=cp_group)
+
+    if cp_rank == 0:
+        # Reconstruct the full logits from CP-split chunks dynamically
+        # Get sequence lengths from cu_seqlens
+        cu_seqlens = batch["cu_seq_lens_q_padded"].cpu()
+        num_seqs = len(cu_seqlens) - 1
+        total_tokens = int(cu_seqlens[-1].item())
+        vocab_size = logits_nondistributed_for_comparison.shape[-1]
+
+        reconstructed_logits = torch.zeros((total_tokens, vocab_size), dtype=torch.bfloat16)
+
+        # For each sequence, reconstruct from CP chunks
+        cp_offset_rank0 = 0
+        cp_offset_rank1 = 0
+
+        for seq_idx in range(num_seqs):
+            seq_start = int(cu_seqlens[seq_idx].item())
+            seq_end = int(cu_seqlens[seq_idx + 1].item())
+            seq_len = seq_end - seq_start
+            chunk_size = seq_len // (2 * cp_world_size)  # Each sequence split into 2*cp_world_size chunks
+
+            # CP rank 0 gets chunks [0, 3], CP rank 1 gets chunks [1, 2]
+            for chunk_idx in range(2 * cp_world_size):
+                chunk_start_in_seq = seq_start + chunk_idx * chunk_size
+                chunk_end_in_seq = chunk_start_in_seq + chunk_size
+
+                if chunk_idx == 0 or chunk_idx == 3:  # Chunks for CP rank 0
+                    reconstructed_logits[chunk_start_in_seq:chunk_end_in_seq, :] = logits_list[0][
+                        cp_offset_rank0 : cp_offset_rank0 + chunk_size, :
+                    ]
+                    cp_offset_rank0 += chunk_size
+                else:  # Chunks 1, 2 for CP rank 1
+                    reconstructed_logits[chunk_start_in_seq:chunk_end_in_seq, :] = logits_list[1][
+                        cp_offset_rank1 : cp_offset_rank1 + chunk_size, :
+                    ]
+                    cp_offset_rank1 += chunk_size
+
+        assert reconstructed_logits.shape == logits_nondistributed_for_comparison.shape
+        torch.testing.assert_close(
+            reconstructed_logits,
+            logits_nondistributed_for_comparison,
+            atol=0.29,
```

Comment:
that's very specific 😟  -- is this just large enough to pass on your machine? can we get a bit of a safety margin for future library + hardware changes?